internal/handlers/common/filter.go

```diff
@@ -1435,3 +1467,53 @@ func filterFieldExprElemMatch(doc *types.Document, filterKey string, exprValue a
 
 	return filterFieldExpr(doc, filterKey, expr)
 }
+
+// filterByComparisonOperandApplicable returns true if comparison operands
+// $gt, $gte, $lt and $lte can be applied to fieldValue and exprValue.
+func filterByComparisonOperandApplicable(fieldValue, exprValue any) bool {
+	if _, ok := fieldValue.(*types.Array); ok {
+		// array is special since it could contain values
+		// which maybe the same type as operand.
+		return true
+	}
+
+	switch exprValue.(type) {
+	case float64, int32, int64:
+		switch fieldValue.(type) {
+		case float64, int32, int64:
+		default:
+			return false
+		}
+	case string:
+		_, ok := fieldValue.(string)
+		return ok
+	case types.Binary:
+		_, ok := fieldValue.(types.Binary)
+		return ok
+	case types.ObjectID:
+		_, ok := fieldValue.(types.ObjectID)
+		return ok
+	case bool:
+		_, ok := fieldValue.(bool)
+		return ok
+	case time.Time:
+		_, ok := fieldValue.(time.Time)
+		return ok
+	case *types.Document:
+		_, ok := fieldValue.(*types.Document)
+		return ok
+	case *types.Array:
+		_, ok := fieldValue.(*types.Array)
+		return ok
+	case types.NullType:
+		_, ok := fieldValue.(types.NullType)
+		return ok
+	case types.Timestamp:
+		_, ok := fieldValue.(types.Timestamp)
+		return ok
+	default:
+		panic("unsupported type")
+	}
+
+	return true
```

Comment:
Let's remove this return as it's hard to follow from which case it's coming. Every case has their own return and then this one is not needed.

Reply:
Removed.